pitop/robotics/navigation/navigation_controller.py

```diff
@@ -0,0 +1,368 @@
+from threading import Thread, Event
+import time
+import numpy as np
+import math
+from typing import Union
+from simple_pid import PID
+import sched
+
+
+class RobotState:
+    def __init__(self):
+        self._x = 0.0      # m
+        self._y = 0.0      # m
+        self._angle = 0.0  # degrees
+        self._theta = 0.0  # radians
+        self.v = 0.0       # m/s
+        self.w = 0.0       # rad/s
+
+    def __str__(self):
+        degree_symbol = u'\N{DEGREE SIGN}'
+        return f"x = {self.x:.3f} m\n" \
+               f"y = {self.y:.3f} m\n" \
+               f"Angle = {self.angle:.3f}{degree_symbol}'\n" \
+               f"Velocity = {self.v:.3f} m/s\n" \
+               f"Angular velocity = {math.degrees(self.w):.3f} {degree_symbol}/s\n" \
+
+
+    @property
+    def x(self):
+        return self._x
+
+    @x.setter
+    def x(self, value):
+        self._x = value
+
+    @property
+    def y(self):
+        return self._y
+
+    @y.setter
+    def y(self, value):
+        self._y = value
+
+    @property
+    def angle(self):
+        return math.degrees(self._theta)
+
+    @angle.setter
+    def angle(self, value):
+        self._theta = math.radians(value)
+
+    @property
+    def theta(self):
+        return self._theta
+
+    @theta.setter
+    def theta(self, value):
+        self._theta = value
+
+    @property
+    def position(self):
+        return self.x, self.y
+
+    @position.setter
+    def position(self, value: tuple):
+        self.x, self.y = value
+
+    def reset_pose(self):
+        self.x, self.y, self.theta = (0.0, 0.0, 0.0)
+
+    def calc_distance(self, point_x, point_y):
+        dx = self.x - point_x
+        dy = self.y - point_y
+        return math.hypot(dx, dy)
+
+
+class GoalCriteria:
+    def __init__(self, linear_speed_factor, angular_speed_factor):
+        self._MAX_GOAL_REACHED_DISTANCE_ERROR = linear_speed_factor * 0.02  # 2 mm at full speed
+        self._MAX_GOAL_REACHED_ANGLE_ERROR = angular_speed_factor * math.radians(4.0)  # 4deg at full speed
+
+    def angle(self, angle_error):
+        return abs(angle_error) < self._MAX_GOAL_REACHED_ANGLE_ERROR:
```

Comment:
```suggestion
        return abs(angle_error) < self._MAX_GOAL_REACHED_ANGLE_ERROR
```

Reply:
How did this get past your syntax checking?! 😝

Reply:
haha I think it was Jorge's suggested commit that I confirmed inside the GitHub web page so didn't catch it